File: Easy/Array/1160.Find-Words-That-Can-Be-Formed-by-Characters.ts
/*

1160. Find Words That Can Be Formed by Characters

You are given an array of strings words and a string chars.

A string is good if it can be formed by characters from chars (each character can only be used once).

Return the sum of lengths of all good strings in words.

 

Example 1:

Input: words = ["cat","bt","hat","tree"], chars = "atach"
Output: 6
Explanation: The strings that can be formed are "cat" and "hat" so the answer is 3 + 3 = 6.
Example 2:

Input: words = ["hello","world","leetcode"], chars = "welldonehoneyr"
Output: 10
Explanation: The strings that can be formed are "hello" and "world" so the answer is 5 + 5 = 10.
 

Constraints:

1 <= words.length <= 1000
1 <= words[i].length, chars.length <= 100
words[i] and chars consist of lowercase English letters.

Topics
Array
Hash Table
String
Counting

*/

function countCharacters(words: string[], chars: string): number {
  // Create a frequency map for the available characters in `chars`
  const charCount = new Map<string, number>();
  for (const ch of chars) {
    charCount.set(ch, (charCount.get(ch) || 0) + 1);
  }

  let totalLength = 0;

  // Helper function to check if the word can be formed
  const canFormWord = (word: string): boolean => {
    const wordCount = new Map<string, number>();
    for (const ch of word) {
      wordCount.set(ch, (wordCount.get(ch) || 0) + 1);
    }

    for (const [ch, count] of wordCount) {
      if ((charCount.get(ch) || 0) < count) {
        return false; // Not enough characters to form the word
      }
    }
    return true; // Word can be formed
  };

  // Iterate through each word in words and check if it can be formed
  for (const word of words) {
    if (canFormWord(word)) {
      totalLength += word.length; // Add the length of the word if it can be formed
    }
  }

  return totalLength;
}

// Example Usage:
console.log(countCharacters(["cat", "bt", "hat", "tree"], "atach")); // Output: 6
console.log(countCharacters(["hello", "world", "leetcode"], "welldonehoneyr")); // Output: 10

File: Easy/Array/1160.Find-Words-That-Can-Be-Formed-by-Characters.go
/*

1160. Find Words That Can Be Formed by Characters

You are given an array of strings words and a string chars.

A string is good if it can be formed by characters from chars (each character can only be used once).

Return the sum of lengths of all good strings in words.



Example 1:

Input: words = ["cat","bt","hat","tree"], chars = "atach"
Output: 6
Explanation: The strings that can be formed are "cat" and "hat" so the answer is 3 + 3 = 6.
Example 2:

Input: words = ["hello","world","leetcode"], chars = "welldonehoneyr"
Output: 10
Explanation: The strings that can be formed are "hello" and "world" so the answer is 5 + 5 = 10.


Constraints:

1 <= words.length <= 1000
1 <= words[i].length, chars.length <= 100
words[i] and chars consist of lowercase English letters.

Topics
Array
Hash Table
String
Counting

⏱ Complexity

Time: O(n * m) — for each word (length m), we check letters

Space: O(1) — only 26 letters in alphabet

*/

package main

import "fmt"

func countCharacters(words []string, chars string) int {
	// Step 1: Count frequency of each character in chars
	charCount := make([]int, 26)
	for _, c := range chars {
		charCount[c-'a']++
	}

	total := 0

	// Step 2: Check each word
	for _, word := range words {
		if canForm(word, charCount) {
			total += len(word)
		}
	}

	return total
}

// Helper function to check if word can be formed
func canForm(word string, charCount []int) bool {
	// Make a copy of charCount to avoid modifying the original
	count := make([]int, 26)
	copy(count, charCount)

	for _, c := range word {
		count[c-'a']--
		if count[c-'a'] < 0 {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(countCharacters([]string{"cat", "bt", "hat", "tree"}, "atach"))            // 6
	fmt.Println(countCharacters([]string{"hello", "world", "leetcode"}, "welldonehoneyr")) // 10
}
